### PR TITLE
Support custom styles for resource style

### LINKED ISSUE #2816: Dashboard "Removed" state still shows green status
We should probably change this to a warning or error status 
image

### PR DESCRIPTION
**Pull request by davidfowl**
- Added support for sending custom style to the dashboard via the resource server protocol.
- Use this for azure resource to show more descriptive state updates.

Contributes to https://github.com/dotnet/aspire/issues/2590
Contributes to https://github.com/dotnet/aspire/issues/2415
Fixes https://github.com/dotnet/aspire/issues/2816

### CODE REVIEW

File: src/Aspire.Hosting/Dashboard/proto/resource_service.proto [hunk 1]
```diff
@@ -155,6 +155,11 @@ message Resource {
 
     // The list of urls that this resource exposes
     repeated Url urls = 13;
+
+    // The style of the state. This is used to determine the state icon.
+    // Supported styles are success", "warning", "info". Any other style
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dashboard/proto/resource_service.proto:160]:
```suggestion
    // Supported styles are "success", "warning", "info". Any other style
```

File: src/Aspire.Hosting/Dashboard/proto/resource_service.proto [hunk 2]
```diff
@@ -155,6 +155,11 @@ message Resource {
 
     // The list of urls that this resource exposes
     repeated Url urls = 13;
+
+    // The style of the state. This is used to determine the state icon.
```
Comment (drewnoakes) [inline src/Aspire.Hosting/Dashboard/proto/resource_service.proto:159]: Did you consider encoding a FluentUI icon name (e.g. `ErrorCircle`), style (e.g. `Filled`) and colour (e.g. `Blue`) here instead? That would give a lot more control over the UI to the server.

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/BicepProvisioner.cs [hunk 3]
```diff
@@ -152,6 +154,13 @@ await foreach (var kv in keyVaults.GetAllAsync(cancellationToken: cancellationTo
 
             if (keyVault is null)
             {
+                await notificationService.PublishUpdateAsync(resource, state => state with
+                {
+                    State = "Provisioning Keyvault",
+                    StateStyle = "info"
+
+                }).ConfigureAwait(false);
```
Comment (drewnoakes) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/BicepProvisioner.cs:162]: To make consistent with the code below:

```suggestion
                })
                .ConfigureAwait(false);
```

**Review by drewnoakes — APPROVED** [review]

File: src/Aspire.Hosting/Dashboard/proto/resource_service.proto (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Dashboard/proto/resource_service.proto:159]: I would prefer that actually. So just parse the enum?

File: src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs [hunk 4]
```diff
@@ -104,11 +98,21 @@ async Task AfterProvisionAsync(IAzureResource resource)
...
+                    StateStyle = "success"
+                })
+                .ConfigureAwait(false);
             }
             catch (Exception)
             {
-                await SetStateAsync(resource, "FailedToStart").ConfigureAwait(false);
+                await UpdateStateAsync(resource, s => s with
+                {
+                    State = "Failed to Provision",
```
Comment (timheuer) [inline src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:112]: later question: these likely need to be localized?

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:112]: Yep.

File: src/Aspire.Hosting/Dashboard/proto/resource_service.proto (hunk 2 above)
Comment (drewnoakes) [reply src/Aspire.Hosting/Dashboard/proto/resource_service.proto:159]: Looks like they're not enums: https://github.com/microsoft/fluentui-blazor/tree/main/src/Assets/FluentUI.Icons/Icons

This method might help, though it doesn't look cheap so results should be cached:

https://github.com/microsoft/fluentui-blazor/blob/main/src/Assets/FluentUI.Icons/Icons.cs#L20

You pass it an instance of [`IconInfo`](https://github.com/microsoft/fluentui-blazor/blob/dev/src/Core/Components/Icons/IconInfo.cs) which takes name, size and variant (filled/regular).

Comment (davidfowl) [reply src/Aspire.Hosting/Dashboard/proto/resource_service.proto:159]: I decided to stick with fixed names.

**Pull request merged**